controllers/usersignup/usersignup_controller.go

```diff
@@ -200,12 +217,191 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
+					r.setStatusMigrationFailedCreate, err, "Failed to migrate UserSignup")
+			}
+		}
+	}
+	return nil
+}
+
+// cleanupMigration performs migration cleanup if the "migration-replaces" annotation has been set on the UserSignup
+//
+// If the annotation has been set, then it indicates that this UserSignup has been migrated, and that the
+// *original* UserSignup should be deleted.  The value of the annotation is the name of the original UserSignup
+// resource that should now be deleted.
+//
+// This function should:
+//
+// 1) Delete the original UserSignup, then
+// 2) Remove the migration-replaces annotation
+func (r *Reconciler) cleanupMigration(userSignup *toolchainv1alpha1.UserSignup, originalUserSignupName string,
+	request ctrl.Request, logger logr.Logger) (ctrl.Result, error) {
+
+	userSignupToDelete := &toolchainv1alpha1.UserSignup{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: request.Namespace,
+		Name:      originalUserSignupName,
+	}, userSignupToDelete)
+	if err == nil {
+		// Delete the original UserSignup, if it has finished deactivating
+		cond, found := condition.FindConditionByType(userSignupToDelete.Status.Conditions, toolchainv1alpha1.UserSignupComplete)
+		if !found || cond.Reason != toolchainv1alpha1.UserSignupUserDeactivatedReason {
+			// The UserSignup to delete isn't finished deactivating yet, return an error
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, userSignup, r.setStatusMigrationFailedCleanup,
+				errs.New("Original UserSignup not deactivated"),
+				fmt.Sprintf("Original UserSignup [%s] not yet deactivated", userSignupToDelete.Name))
+		}
+
+		err = r.Client.Delete(context.TODO(), userSignupToDelete)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, userSignup, r.setStatusMigrationFailedCleanup,
+				err, fmt.Sprintf("Failed to remove original UserSignup [%s]",
+					userSignup.Annotations[migrationAnnotationName]))
+		}
+		// Requeue so that the annotation will now be removed also
+		return reconcile.Result{Requeue: true, RequeueAfter: time.Second}, nil
+	}
+
+	if err != nil {
+		// If the annotation exists however the original UserSignup isn't found, then remove the annotation and the
+		// migration condition from the status
+		if errors.IsNotFound(err) {
+			// Remove the annotation
+			delete(userSignup.Annotations, migrationAnnotationName)
+
+			// Remove the migration condition from the status
+			conditions := []toolchainv1alpha1.Condition{}
+			for _, cond := range userSignup.Status.Conditions {
+				if cond.Type != UserMigrationFailed {
+					conditions = append(conditions, cond)
+				}
+			}
+			userSignup.Status.Conditions = conditions
+
+			// Update the UserSignup
+			err = r.Client.Update(context.TODO(), userSignup)
+			if err != nil {
+				return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, userSignup,
+					r.setStatusMigrationFailedCleanup, err, "Failed to remove migration annotation")
+			}
```

Comment:
btw, do we need to remove the annotation and the condition at all? would it be a big deal if it stays there?
I would keep it there - can be useful for monitoring purposes - to see what has been migrated and what has not.

Reply:
I've got no big problem leaving it there.